### ISSUE TITLE
proposal: x/net/bosh: New package

### ISSUE THREAD

**Issue by SamWhited** [golang/go#14726]
I'd like to propose a new package at `x/net/bosh` that would implement an API similar to [`x/net/websocket`](https://godoc.org/golang.org/x/net/websocket) except for [XEP-0124: Bidirectional-streams Over Synchronous HTTP (BOSH)](https://xmpp.org/extensions/xep-0124.html).

BOSH is an HTTP long-polling technique for emulating a bidirectional text stream (mostly XML, but the spec leaves it open ended so that you can stream anything) that was originally created (and is most often used for) XMPP via [XEP-0206: XMPP Over Bosh](https://xmpp.org/extensions/xep-0206.html).

This proposal is to implement a simple HTTP handler for BOSH, and ensure that the API can be used to implement profiles such as the XMPP over bosh spec (though this probably wouldn't be implemented in the package itself).

I have started on an example implementation [here](https://godoc.org/bitbucket.org/mellium/bosh), but I think the actual BOSH part of it might be a good fit for the x/net/ packages (and keeping it right alongside the websocket implementation would be nice for servers that want to support both transports).

The API I'm proposing to start is just a simple `http.Handler`:

```go
type SessionManager

// Cleanly shut down all sessions managed by the session manager.
func (*SessionManager) Close() {}

// Returns a channel that is closed when the session manager shuts down.
func (*SessionManager) Done() <-chan bool {}

// Satisfy http.Handler and handle incoming BOSH payloads.
func (sm *SessionManager) ServeHTTP(w http.ResponseWriter, r *http.Request) {}
```

**Comment by bradfitz** [comment]
Does anything use this but XMPP? It might make more sense in /x/ if we had an XMPP implementation in /x/.

You would also have to make sure there are enough other people available to be reviewers if it lived in /x/. I at least am not available at the moment to review. I'm not sure anybody else is, either. This might be too specific to stir people's interest. But that's just my guess.

**Comment by SamWhited** [comment]
> Does anything use this but XMPP?

I know there are some video games that use it; eg. according to these [blog](https://metajack.im/2008/09/08/which-bosh-server-do-you-need/) [posts](https://metajack.im/2008/07/02/xmpp-is-better-with-bosh/ there is (was?) a Chess game using it at one point, but I'm having a hard time searching for "things that aren't XMPP that use BOSH". In general, it's mostly XMPP, both chat and internet-of-things devices that need to make it through most firewalls.

I'm pretty sure I've read that some of the other large non-XMPP chat providers were using it, and I know that I've read about various mobile apps using it in environments where it's hard to keep persistant TCP connections open (for things that weren't XMPP), but I'm having trouble finding references for that.

Also, lots of sites provide long-polling APIs that don't use BOSH, but do their own non-standard implementations. BOSH can be thought of as a way of standardizing long-polling, so having an implementation might help make long-polling APIs developed with Go more consistant. I'll see what else I can find.

TL;DR — It's definitely most popular in XMPP, but it's definitely used elsewhere, if sparingly.

> It might make more sense in /x/ if we had an XMPP implementation in /x/.

I'd still love to have that one day and think XMPP is widely applicable outside the instant messaging space as a signaling layer, but that's a discussion for a future proposal.

> You would also have to make sure there are enough other people available to be reviewers if it lived in /x/.

Reviewers is always a tricky one; I'm not sure what to do about that. It seems like the same people who review websockets commits would potentially also be good candidates to review BOSH commits when / if they have the time?

**Comment by bradfitz** [comment]
I think this is probably developed best out of tree. There's little to be gained to putting it in /x/, but a lot of burden.

As an aside, one concrete API suggestion: use `Done() <-chan struct{}` instead of `chan bool`. That's the normal Go pattern these days (e.g. `Context.Done`, `Request.Cancel`, `Dialer.Cancel`, ignoring the old mistake of `CloseNotifier` which used a `bool`).

We can revisit this if we find reviewers and/or BOSH becomes an increasingly important dependency for other packages.
